### ISSUE TITLE
Problem with bigger heterozygous regions

### ISSUE THREAD

**Issue by dominik-handler** [nanoporetech/medaka#121]
Hi,

I tried polishing a genome with quite many heterozygous transposon insertions. I previously used medaka already (v0.7) without any problem. Now with the latest version, I see that medaka starts to insert "random" bases. As far as I can tell this happens at positions where a heterozygous transposon is present in a fraction of the reads but not in the primary assembly. Typically I get stretches of homopolymer or low-complexity sequences inserted. In the 4x Racon polished sequence no insertion is present.

Do you know how I can avoid this behavior? 

All the best, Dominik

![image](https://user-images.githubusercontent.com/12491961/71898280-32a8e000-3159-11ea-9812-2a13488743c8.png)

**Comment by cjw85** [comment]
Hi @dominik-handler 

I think the clue to understanding this might be to view the alignments in the intermediate `.bam` file that medaka produces in its working directory, and either post-process them or change the alignment parameters used.

It's not an assumption per se, but medaka's algorithm is somewhat built to process homozygous sequences (`medaka_consensus` is certainly only built to a single linear sequence). It might be that the best solution is to find a way to remove reads with the additional transposon sequence.

The code snippet [here](https://nanoporetech.github.io/medaka/installation.html#improving-parallelism) shows how to run the core medaka program where the user directly provides alignments.

I'm curious that version 0.7 gave results on this data but that the current version gives the above. Is it possible that you could share your data to help understand this issue?

**Comment by dominik-handler** [comment]
Cool, I will have a look. For the 0.7 version, I will redo the polishing on the current assembly with 0.7 to compare. The difference between my polishing runs has been different primary assemblies with the same read-data. Also, the old primary assembly was missing many heterozygous transposons. Is there a way that I can use the current Guppy models with 0.7?

Do you think it is possible to use this as a feature to incorporate the TE sequences into the genome? I tried to re-polish the genome version with the random-sequences inserted without good results. I always got the contig split in these regions. Probably minimap was not able to align the reads through these regions and created a split alignment. Do you think medaka can create the correct sequence if I manage to get the reads aligned through by changing minimap settings?

Dominik

**Comment by dominik-handler** [comment]
Hi, I experimented a bit. I used Guppy 3.4.1 basecalled reads, trimmed and filtered them for quality. These reads I used for assembly and polishing with Racon and Medaka. The primary assembly was done using MARVEL and I polished this genome 4x with Racon. This polished assembly and the reads used for assembly are used for all polishing tests and the only variable is the Medaka version.

After polishing I get the following genome sizes:
Racon polished --> 157.082 Mb
Medaka 0.7.1 --> 96.3283 Mb (or 99.3078 in a 2nd run)
Medaka 0.11.3 --> 155.965 Mb
Medaka 0.11.3 using the Medaka 0.7.1 bam --> 158.581 Mb

Seems Medaka 0.7.1 did not like the newer Guppy reads but I also cannot see any induced homopolymers. Also 0.7 did not introduce them when used with older basecalls and another primary assembly (yet the same raw nanopore data). Medaka 0.11.3 introduced the stretches inin both tested cases. 

How can I best send you my data? 

Dominik

**Comment by cjw85** [comment]
This observation is interesting:

    Medaka 0.11.3 using the Medaka 0.7.1 bam --> 158.581 Mb

Looking through the code [history](https://github.com/nanoporetech/medaka/commit/f6e8f2e972e7a4105c32d4dd571388780d215544) I see that between v0.7.0 and v0.11.3 we changed the version of the `minimap2` aligner that we use when installing medaka from source and via bioconda. This may explain some of the differences.

Another difference may be that in [v0.9.0](https://github.com/nanoporetech/medaka/releases/tag/v0.9.0) we modified some of medaka's logic of how contigs are split when the input bam does not contain spanning reads. Currently the behaviour is that contigs are split in this scenario; these splits are annotated in the header of the output fasta so in principle is it possible to reference medaka's output back to the input sequence.

A point of note that might be relevant here is that medaka only uses primary alignments from the input bam so it may be that even if your bam contains sensible secondary/supplementary alignments, medaka is ignoring them.

We can try to get your data to us, but we can't promise to take a look. Nevertheless I hope that the comments above are helpful to you in understanding medaka's behaviour.
